### ISSUE TITLE
[Feature Request] Font Selection + Letter/Icon/Image (Individual size settings)

### ISSUE THREAD

**Issue by rezpower** [zombieFox/nightTab#196]
This extension is already a killer and getting better every day. The implementation of Icons was really great! (Thanks for this) :)

I have few things in mind that I wanted to share:

1- For now I am getting my icons from api.faviconkit.com and s2.googleusercontent.com/s2/favicons?domain=X
Some icons are also direct links to an image on the site. But it happens that some are too small and some Big. I was wondering if it would be possible to also add Letter/Icon/Image size controls to individual bookmarks instead of general setting that affects all bookmarks at the same time? Same thing with the letter (if we could individually set a size).

2- It would be nice also to be able to choose fonts for the letters (Individually if possible). For some bookmarks that do not have an icon, this could allow  nicer customization.

3- We can choose custom colors individually for theme and accents. It would be nice to have the option also for Names.

I know and it's been explained to me that there is restriction for having the icons stored locally. I wish it was possible but I understand it's not. But do you guys know any service where we can upload all our icons, so we can add url to that site and pull all of them down to nighttab? I tried to do this with my google drive, dropbox and even imagur and it didn't work!

**Comment by Excigma** [comment]
You can upload the images to a Github repo. Imgur URLs should work though??

**Comment by zombieFox** [comment]
> This extension is already a killer and getting better every day. The implementation of Icons was really great!

Thank you

> But it happens that some are too small and some Big. I was wondering if it would be possible to also add Letter/Icon/Image size controls to individual bookmarks instead of general setting that affects all bookmarks at the same time? Same thing with the letter (if we could individually set a size).

To be clear, as I anticipate others will experience different image sizes (so this is more for documentation and may end up in the readme), let's run through the baseline for the Bookmark Image feature:

Each bookmark has a display area, (black box). Inside is a letter/icon/image box (red box) and a name box (blue box). 

![Screenshot 2020-06-24 at 08 26 28](https://user-images.githubusercontent.com/7115017/85515511-1e687a80-b5f5-11ea-8aaa-f90ab83374b3.png)

The letter/icon/image box is sized the same whether displaying a letter, icon or image and can be scaled up or down using controls in the menu. When scaled up or down they keep the same square shape.

When developing the Bookmark image feature one of the expectations was to release some additional control and responsibility about the experience to the end user. So when a user adds an image as the letter/icon/image element they will have to make sure the image is cropped to show the maximum amount of image. This will ensure the best fit inside the image box (red box).

![Screenshot 2020-06-24 at 08 24 07](https://user-images.githubusercontent.com/7115017/85516640-57551f00-b5f6-11ea-9ee1-4a85cc815a0b.png)

Looking at the example above the image is not cropped exactly to the edges (the round logo edge is not touching the red box which shows us the image bounds). So shows up a bit smaller than the letters or icons on a bookmark tile.

Adding individual bookmark size controls for either letter, icon and images is difficult and I estimate a large task. I'll investigate but I don't think I can commit to it right now. 

> 2- It would be nice also to be able to choose fonts for the letters (Individually if possible). For some bookmarks that do not have an icon, this could allow nicer customization.

This is similar, would be a nice additional feature but estimated as a large task to develop. 

> 3- We can choose custom colors individually for theme and accents. It would be nice to have the option also for Names.

Could be, I saw this coming as the background and accent colour for each bookmark tile can be changed. I'll see what I can do.

> But do you guys know any service where we can upload all our icons, so we can add url to that site and pull all of them down to nighttab? I tried to do this with my google drive, dropbox and even imagur and it didn't work!

I would recommend what @Excigma has said, upload to a github repo. They are free and linking to images from them has worked for me just fine. 

👍

**Comment by rezpower** [comment]
Thanks a lot for the detailed explanation . The GitHub tip was already a game changer as it's now possible to have a storage, create customized icons etc... 
For the icon sizes I totally understand the difficulty and maybe after all not worth the effort if it's too hard to implement. The reason I asked this was to be able to somehow break the aspect ratio rule when using square icons vs rectangulaire ones. (see examples ):
![Annotation 2020-06-24 155106](https://user-images.githubusercontent.com/28585954/85569876-34952b80-b633-11ea-82bd-f0019dd4d496.png)
![Annotation 2020-06-24 155150](https://user-images.githubusercontent.com/28585954/85569891-3828b280-b633-11ea-9cfb-944c038a685a.png)

For my 3rd request (Choosing custom colors individually for Names) After thinking about it a little I think that maybe another solutions would be to have a option under Name: **_Use accent color_**  (talking about the accent override color of course)!

**Comment by zombieFox** [comment]
> For the icon sizes I totally understand the difficulty and maybe after all not worth the effort if it's too hard to implement. The reason I asked this was to be able to somehow break the aspect ratio rule when using square icons vs rectangulaire ones.

I've doing some investigating on how much work exactly it might be. _Spoilers:_ yup, large amounts of refactoring. 

> For my 3rd request (Choosing custom colors individually for Names) After thinking about it a little I think that maybe another solutions would be to have a option under Name: Use accent color (talking about the accent override color of course)!

This could happen. The colour of the name is defined by the theme, so maybe another one of these modules would work in the future:

<img width="475" alt="Screenshot 2020-06-26 at 08 19 28" src="https://user-images.githubusercontent.com/7115017/85831175-cc0a9380-b785-11ea-9b39-b48ae1318c74.png">

Will continue to investigate.

**Comment by zombieFox** [comment]
On the plus side the investigation lead to a new feature I've been thinking of for some time. Any of the above requests would be blocked by this feature so I thought why not see how much effort it would be, seems not a lot. Check out [the bookmark preview PR.](https://github.com/zombieFox/nightTab/pull/199#issue-439946769)

**Comment by zombieFox** [comment]
**WIP** feature PR: add link item advanced controls #202


New individual bookmarks controls for:

- TIle background image and opacity
- Wide tile (Bookmark tile to span across two columns.)
- Tall tile (Bookmark tile to span across two rows.)
- Letter size
- Icon size
- Image size
- Letter or Icon drop shadow
- Name size
- Visual element and Name alignment, Rotation, Horizontally offset, Vertically offset
- Gutter spacing
- Vertically or Horizontally alignment
- Visual element and Name order
- Colour
- Accent

![Kapture 2020-07-17 at 16 35 36](https://user-images.githubusercontent.com/7115017/87804578-f9f66b80-c84b-11ea-9fa8-1b89b667349f.gif)


![Kapture 2020-07-17 at 16 37 43](https://user-images.githubusercontent.com/7115017/87804561-f367f400-c84b-11ea-8d10-56388393e3f4.gif)

![form](https://user-images.githubusercontent.com/7115017/87805293-ff07ea80-c84c-11ea-92cd-f32f7bb39687.png)

**Comment by zombieFox** [comment]
Existing customisation controls for bookmarks have migrated to the bookmark level in [v6.1.1](https://github.com/zombieFox/nightTab/releases/tag/v6.1.1). This is far as I wish to take the refactoring.
